plugins/inputs/tacacs/tacacs.go

```diff
@@ -0,0 +1,208 @@
+		RemAddr:       t.RequestAddr,
+	}
+
+	return nil
+}
+
+func (t *Tacacs) AuthenReplyToString(code uint8) string {
+	switch code {
+	case tacplus.AuthenStatusPass:
+		return `AuthenStatusPass`
+	case tacplus.AuthenStatusFail:
+		return `AuthenStatusFail`
+	case tacplus.AuthenStatusGetData:
+		return `AuthenStatusGetData`
+	case tacplus.AuthenStatusGetUser:
+		return `AuthenStatusGetUser`
+	case tacplus.AuthenStatusGetPass:
+		return `AuthenStatusGetPass`
+	case tacplus.AuthenStatusRestart:
+		return `AuthenStatusRestart`
+	case tacplus.AuthenStatusError:
+		return `AuthenStatusError`
+	case tacplus.AuthenStatusFollow:
+		return `AuthenStatusFollow`
+	}
+	return "AuthenStatusUnknown(" + strconv.FormatUint(uint64(code), 10) + ")"
+}
+
+func (t *Tacacs) Gather(acc telegraf.Accumulator) error {
+	var wg sync.WaitGroup
+
+	for idx := range t.clients {
+		wg.Add(1)
+		go func(client *tacplus.Client) {
+			defer wg.Done()
+			acc.AddError(t.pollServer(acc, client))
+		}(&t.clients[idx])
+	}
+
+	wg.Wait()
+	return nil
+}
+
+func (t *Tacacs) pollServer(acc telegraf.Accumulator, client *tacplus.Client) error {
+	// Create the fields for this metric
+	tags := map[string]string{"source": client.Addr}
+	fields := make(map[string]interface{})
+
+	secret, err := t.Secret.Get()
+	if err != nil {
+		return fmt.Errorf("getting secret failed: %w", err)
+	}
+	defer config.ReleaseSecret(secret)
+
+	client.ConnConfig.Secret = secret
+
+	username, err := t.Username.Get()
+	if err != nil {
+		return fmt.Errorf("getting username failed: %w", err)
+	}
+	defer config.ReleaseSecret(username)
+
+	password, err := t.Password.Get()
+	if err != nil {
+		return fmt.Errorf("getting password failed: %w", err)
+	}
+	defer config.ReleaseSecret(password)
+
+	ctx := context.Background()
+	if t.ResponseTimeout > 0 {
+		var cancel context.CancelFunc
+		ctx, cancel = context.WithTimeout(ctx, time.Duration(t.ResponseTimeout))
+		defer cancel()
+	}
+
+	startTime := time.Now()
+	reply, session, err := client.SendAuthenStart(ctx, &t.authStart)
+	if err != nil {
+		if !errors.Is(err, context.DeadlineExceeded) {
+			return fmt.Errorf("error on new tacacs authentication start request to %s : %w", client.Addr, err)
+		}
+		fields["responsetime_ms"] = time.Duration(t.ResponseTimeout).Milliseconds()
+		fields["response_status"] = "Timeout"
+		acc.AddFields("tacacs", fields, tags)
+		return nil
+	}
+	defer session.Close()
+	if reply.Status != tacplus.AuthenStatusGetUser {
+		fields["responsetime_ms"] = time.Since(startTime).Milliseconds()
+		fields["response_status"] = t.AuthenReplyToString(reply.Status)
+		acc.AddFields("tacacs", fields, tags)
+		return nil
+	}
+
+	reply, err = session.Continue(ctx, string(username))
+	if err != nil {
+		if !errors.Is(err, context.DeadlineExceeded) {
+			return fmt.Errorf("error on tacacs authentication continue username request to %s : %w", client.Addr, err)
+		}
+		fields["responsetime_ms"] = time.Duration(t.ResponseTimeout).Milliseconds()
```

Comment:
I think that is a valid point, especially for users who would set the timeout to zero. In other cases, it will be close to the configured timeout value anyway. With zero timeout configured it can theoretically also timeout from other sources than the configured context deadline (from some network functions called down the line). Maybe in addition to context.DealineExceeded error we should also allow errors.Is(err, os.ErrDeadlineExceeded) for these cases.

Reply:
While the used tacacs package seems to be correctly setting the timeout to the deeper used network calls, I could not deterministically say that it will always return only context.DeadlineExceeded error on timeout. 

It Seemd os.IsTimeout() is would be better here, as it returns True for both context.DeadlineExceeded and os.ErrDeadlineExceeded and possibly other timeout time errors, but after testing it it was too lenient/broad and it was returning True even for immediate dial tcp or dns errors etc so will keep 
`if !errors.Is(err, context.DeadlineExceeded) && !errors.Is(err, os.ErrDeadlineExceeded)`

The timeout time returned is now real value.